Document builtin topics and blocklist [13993] (#154)
eProsima/DDS-Router: docs/rst/user_manual/configuration.rst

This *.yaml* file contains all the information regarding the |ddsrouter| configuration, such as topics filtering
and :term:`Participants <Participant>` configurations.

.. _topic_filtering:

Topic Filtering
===============

|ddsrouter| allows filtering of DDS :term:`Topics<Topic>`, that is, it allows to define which DDS Topics are going to be
relayed by the application.
This way, it is possible to define a set of rules in |ddsrouter| to filter those data samples the user does not wish to
forward. It is not mandatory to define such set of rules in the configuration file.
In this case, a DDS Router will forward all the data published under the topics that it automatically discovers within

> the DDS network to which it connects.

Reviewer: I will try to refactor these sentences.
First, separate the explanation about the configuration and the use of topic filtering.
Second, saying that data samples are filtered is incorrect, it should say DDS entities are created/activated in those topics.
Third, make it clearer that when a new topic is discovered in the network, Router will create entities to forward it if allowed.

Author: The use of these phrases without reference to the creation of internal entities is intended to be more understandable to the user. Therefore I would leave them as they are.

Topic Filtering
===============

> |ddsrouter| allows filtering of DDS :term:`Topics<Topic>`, that is, it allows defining which DDS Topics are going to be

Reviewer 2: ```suggestion
|ddsrouter| allows filtering of DDS :term:`Topics<Topic>`, that is, it allows to define which DDS Topics are going to be
```

relayed by the application.

> This way, it is possible to define a set of rules in |ddsrouter| to filter those data that the user do not want to be
> forwarded. It is not mandatory to define a rule set in the configuration file.

Reviewer 2: ```suggestion
This way, it is possible to define a set of rules in |ddsrouter| to filter those data samples the user does not wish to forward. It is not mandatory to define such set of rules in the configuration file.
```

> In this case, a DDS Router will forward all the data published under the topics that it automatically discovers within
> the DDS network to which it connects.

Reviewer 2: ```suggestion
In this case, a DDS Router will forward all the data published under the topics that it automatically discovers within
the DDS network to which it is connected.
```

To define these data filtering rules based on the Topics to which they belong, three lists are available:

* Builtin topics list (``builtin-topics``)
* Allowed topics list (``allowlist``)

> * Block topics list (``blocklist``)

Reviewer 2: Maybe better explain `builtin-topics` the last one, and let's try to use this ordering everywhere else.

    *   - ``keyed``
        - ``bool``
        - Both ``true`` and ``false``

The entry ``keyed`` determines whether the corresponding topic is `keyed <https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/typeSupport/typeSupport.html#data-types-with-a-key>`_
or not. See :term:`Topic` section for further information about the topic.

.. note::

    Tags ``allowlist``, ``blocklist`` and ``builtin-topics`` must be at yaml base level (it must not be inside any
    other tag).

.. note::

    Placing quotation marks around values in a YAML file is generally optional. However, values containing wildcard
    characters must be enclosed by single or double quotation marks.

Allow topic list (``allowlist``)
--------------------------------
This is the list of topics that |ddsrouter| will forward, i.e. the data published under the topics matching the
expressions in the ``allowlist`` will be relayed by |ddsrouter|.

> 

Reviewer: Add a note (or warning) saying that an empty allowlist allow every topic

Block topic list (``blocklist``)
--------------------------------
This is the list of topics that the |ddsrouter| will block, that is, all data published under the topics matching the
filters specified in the ``blocklist`` will be discarded by the |ddsrouter| and therefore will not be relayed.

This list takes precedence over the ``allowlist``.
If a topic matches an expression both in the ``allowlist`` and in the ``blocklist``, the ``blocklist`` takes precedence,
causing the data under this topic to be discarded.

Builtin topics list (``builtin-topics``)
----------------------------------------

|ddsrouter| includes a mechanism to automatically detect which topics are being used in a DDS network.
By automatically detecting these topics, a |ddsrouter| creates internal DDS :term:`Writers<DataWriter>`
and :term:`Readers<DataReader>` for each topic and for each Participant in order to relay the data published on each

> discovered topic.

Reviewer: I would move this to the main Topic filtering paragraph, I think it is main information.
If not this, at least mention it above

Reviewer 2: It is actually in the first paragraph.

The entry ``keyed`` determines whether the corresponding topic is `keyed <https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/typeSupport/typeSupport.html#data-types-with-a-key>`_
or not. See Topic section for further information about the topic.

.. note::

    Tags ``allowlist``, ``blocklist`` and ``builtin-topics`` must be at yaml base level (it must not be inside any
    other tag).

Allow topic list (``allowlist``)
--------------------------------

> This is the list of topics that |ddsrouter| will forward, i.e. the data published under the topics defined in
> the ``allowlist`` will be relayed by |ddsrouter|.

Reviewer 2: ```suggestion
This is the list of topics that |ddsrouter| will forward, i.e. the data published under the topics matching the expressions in
the ``allowlist`` will be relayed by |ddsrouter|.
```

Block topic list (``blocklist``)
--------------------------------

> This is the list of topics that the |ddsrouter| will block, that is, all data published under the topics defined in
> the ``blocklist`` will be discarded by the |ddsrouter| and therefore will not be relayed.

Reviewer 2: ```suggestion
This is the list of topics that the |ddsrouter| will block, that is, all data published under the topics matching the expressions in
the ``blocklist`` will be discarded by the |ddsrouter| and therefore will not be relayed.
```

This list takes precedence over the ``allowlist``.

> If there is a topic defined in the ``allowlist`` and in the ``blocklist``, the ``blocklist`` takes precedence, causing
> the data of this topic to be discarded.

Reviewer 2: ```suggestion
If a topic matches an expression both in the ``allowlist`` and in the ``blocklist``, the ``blocklist`` takes precedence, causing
the data under this topic to be discarded.
```

Builtin topics list (``builtin-topics``)
----------------------------------------

> |ddsrouter| has a mechanism to automatically detect which topics are being used in a DDS network.

Reviewer 2: ```suggestion
|ddsrouter| includes a mechanism to automatically detect which topics are being used in a DDS network.
```

By automatically detecting these topics, a |ddsrouter| creates internal DDS :term:`Writers<DataWriter>`
and :term:`Readers<DataReader>` for each topic and for each Participant in order to relay the data published on each
discovered topic.

The discovery phase of the network topics can be accelerated by using the builtin topic list (``builtin-topics``).

> By defining topics in this list, the DDS router will create the DataWriters and DataReaders for these topics without
> the need to have previously discovered them.

Reviewer 2: ```suggestion
By defining topics in this list, the DDS router will create the DataWriters and DataReaders for these topics without waiting for them to be discovered.
```

In this way, the initialization phase mentioned above is omitted and the application launching efficiency is improved.

Examples of usage
-----------------

The following is an example of how to use the ``allowlist``, ``blocklist`` and ``builtin-topics`` configurations to
setup the |ddsrouter| filtering rules.

Dynamic topic discovery example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This example shows how the |ddsrouter| is initially configured to forward the ``rt/chatter`` topic (default ROS 2
topic for ``talker`` and ``listener``) with type name ``std_msgs::msg::dds_::String_``, while the rest of the
topics in the DDS network are expected to be dynamically discovered.
Additionally, a rule is defined to block only the ``HelloWorldTopic`` topic.

.. code-block:: yaml

    blocklist:

>       - name: "HelloWorldTopic"

Reviewer 2: Remove quotes everywhere.

>       - name: "HelloWorldTopic"

Reviewer 2: ```suggestion
      - name: *
```

Reviewer 2: Just a suggestion in order to include wildcard expressions in the example, maybe it is more interesting for an user to block all topics of a certain type.

        type: "HelloWorld"

    builtin-topics:
      - name: "rt/chatter"
        type: "std_msgs::msg::dds_::String_"

Allowlist and blocklist collision
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In the following example, the ``HelloWorldTopic`` topic is in both the ``allowlist`` and the

> ``blocklist`` so according to the ``blocklist`` preference rule, this topic is blocked.

Reviewer 2: ```suggestion
``blocklist``,  so according to the ``blocklist`` preference rule this topic is blocked.
```

      - name: HelloWorldTopic
        type: HelloWorld

    blocklist:

>       - name: "*"
>         type: HelloWorld

Reviewer: Use real cases to exemplify. rq and rr are the name for topics in services in ROS2 
```suggestion
      - name: "rq/*"
      - name: "rr/*"
```

Reviewer: The idea was to filter them, so the whole .yaml refers to a real use case.
